api/src/main/java/com/inrupt/client/ProblemDetails.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package com.inrupt.client;
+
+import com.inrupt.client.spi.JsonService;
+import com.inrupt.client.spi.ServiceProvider;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Optional;
+
+/**
+ * A data class representing a structured problem description sent by the server on error response.
+ *
+ * @see <a href="https://www.rfc-editor.org/rfc/rfc9457">RFC 9457 Problem Details for HTTP APIs</a>
+ */
+public class ProblemDetails {
+    public static final String MIME_TYPE = "application/problem+json";
+    public static final String DEFAULT_TYPE = "about:blank";
+    private final URI type;
+    private final String title;
+    private final String details;
+    private final int status;
+    private final URI instance;
+    private static JsonService jsonService;
+    private static boolean isJsonServiceInitialized;
+
+    public ProblemDetails(
+        final URI type,
+        final String title,
+        final String details,
+        final int status,
+        final URI instance
+    ) {
+        this.type = type;
+        this.title = title;
+        this.details = details;
+        this.status = status;
+        this.instance = instance;
+    }
+
+    public URI getType() {
+        return this.type;
+    };
+
+    public String getTitle() {
+        return this.title;
+    };
+
+    public String getDetails() {
+        return this.details;
+    };
+
+    public int getStatus() {
+        return this.status;
+    };
+
+    public URI getInstance() {
+        return this.instance;
+    };
+
+    public static class ProblemDetailsData {
+        public URI type;
+        public String title;
+        public String details;
+        public int status;
+        public URI instance;
+    }
+
+    private static JsonService getJsonService() {
```

Comment:
I thought it would be nicer, but after having implemented some of it, it looks quite the same. One of my assumptions which causes the code to look like what it does here is that the SPI lookup is a relatively expensive operation that you only want to do once, so I want to have a clear separation between "this is null because we haven't initialized it" and "this is null because the lookup failed", so that I don't do a systematic lookup in the latter case assuming we might be in the former.

If I have understood what you are describing correctly, having the `Optional` doesn't change much to the overall structure, in particular because a lookup failure results in an exception, and not in null, so the try/catch needs to be there even when initializing the `Optional` (I think)